mlir/lib/Dialect/Affine/Utils/LoopFusionUtils.cpp

```diff
@@ -275,7 +293,13 @@ FusionResult mlir::affine::canFuseLoops(AffineForOp srcForOp,
   // 'forOpA' executes before 'forOpB' in 'block'.
   auto forOpA = isSrcForOpBeforeDstForOp ? srcForOp : dstForOp;
   auto forOpB = isSrcForOpBeforeDstForOp ? dstForOp : srcForOp;
-
+  // Check if either of the loop has any possible side-effects.
+  // if yes then fusion precondition failed.
+  if (hasPossibleSideEffects(forOpA) || hasPossibleSideEffects(forOpB)) {
```

Comment:
This isn't properly framed. You need to check for side-effecting ops other than affine ops that have already been accounted for (affine.load/store/alloc,dealloc). So this should be called `hasNonAffineSideEffects` and the latter should in turn use `isMemoryEffectFree` while excepting affine ops accounted for. In fact, this method isn't the right place to check for this. The MDG construction itself (mdg->init()) itself has checks where we are missing a check for unknown ops.